runtime/op/join/join.go

```diff
@@ -33,11 +34,25 @@ type Op struct {
 	types       map[int]map[int]*zed.TypeRecord
 }
 
-func New(octx *op.Context, anti, inner bool, left, right zbuf.Puller, leftKey, rightKey expr.Evaluator, leftOrder, rightOrder order.Which, lhs field.List, rhs []expr.Evaluator) (*Op, error) {
+func New(octx *op.Context, anti, inner bool, left, right zbuf.Puller, leftKey, rightKey expr.Evaluator,
+	leftOrder, rightOrder order.Which, leftSorted, rightSorted bool, lhs field.List,
+	rhs []expr.Evaluator) (*Op, error) {
 	cutter, err := expr.NewCutter(octx.Zctx, lhs, rhs)
 	if err != nil {
 		return nil, err
 	}
+	if !leftSorted {
+		left, err = sort.New(octx, left, []expr.Evaluator{leftKey}, leftOrder, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if !rightSorted || leftOrder != rightOrder {
+		right, err = sort.New(octx, right, []expr.Evaluator{rightKey}, rightOrder, true)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
Something's wrong here.

Suppose you create two files.
```console
$ cat <<EOF >a.zson
{a:1}
{a:2}
EOF
$ cat <<EOF >b.zson
{b:2}
{b:1}
EOF
```

Joining those files on `a=b` should always produce two output values.  But in these cases it doesn't.
```console
$ zq 'file a.zson | join (file b.zson | sort -r b) on a=b'
{a:2}
$ zq 'file a.zson | sort a | join (file b.zson | sort -r b) on a=b'
{a:2}
$ zq 'file a.zson | sort -r a | join (file b.zson) on a=b'
{a:1}
$ zq 'file a.zson | sort -r a | join (file b.zson | sort b) on a=b'
{a:1}
```

Reply:
This is v strange, thanks for digging into this.

Reply:
Actually a much better test as well will use this.